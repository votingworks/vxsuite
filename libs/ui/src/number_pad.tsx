import React, { useCallback, useEffect, useRef } from 'react';
import styled from 'styled-components';
import { Button } from './button';

export const NumberPadContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  > button {
    display: flex;
    justify-content: center;
    margin: 2px;
    width: 26%;
  }
`;

const DIGITS = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
export interface NumberPadProps {
  onButtonPress: (buttonValue: number) => void;
  onBackspace: () => void;
  onClear: () => void;
  onEnter?: () => void;
}

export function NumberPad({
  onButtonPress,
  onBackspace,
  onClear,
  onEnter,
}: NumberPadProps): JSX.Element {
  const container = useRef<HTMLDivElement>(null);
  const onKeyPress: React.KeyboardEventHandler = useCallback(
    (event) => {
      if (DIGITS.includes(event.key)) {
        // eslint-disable-next-line vx/gts-safe-number-parse
        onButtonPress(Number(event.key));
      } else if (event.key === 'x') {
        onClear();
      } else if (onEnter !== undefined && event.key === 'Enter') {
        onEnter();
      }
    },
    [onButtonPress, onClear, onEnter]
  );
  const onKeyDown: React.KeyboardEventHandler = useCallback(
    (event) => {
      if (event.key === 'Backspace') {
        onBackspace();
      }
    },
    [onBackspace]
  );

  useEffect(() => {
    /* istanbul ignore next */
    container.current?.focus();
  }, []);

  return (
    <NumberPadContainer
      tabIndex={0}
      ref={container}
      onKeyPress={onKeyPress}
      onKeyDown={onKeyDown}
    >
      <Button onPress={useCallback(() => onButtonPress(1), [onButtonPress])}>
        1
      </Button>
      <Button onPress={useCallback(() => onButtonPress(2), [onButtonPress])}>
        2
      </Button>
      <Button onPress={useCallback(() => onButtonPress(3), [onButtonPress])}>
        3
      </Button>
      <Button onPress={useCallback(() => onButtonPress(4), [onButtonPress])}>
        4
      </Button>
      <Button onPress={useCallback(() => onButtonPress(5), [onButtonPress])}>
        5
      </Button>
      <Button onPress={useCallback(() => onButtonPress(6), [onButtonPress])}>
        6
      </Button>
      <Button onPress={useCallback(() => onButtonPress(7), [onButtonPress])}>
        7
      </Button>
      <Button onPress={useCallback(() => onButtonPress(8), [onButtonPress])}>
        8
      </Button>
      <Button onPress={useCallback(() => onButtonPress(9), [onButtonPress])}>
        9
      </Button>
      <Button onPress={onClear}>
        <span role="img" aria-label="clear">
          ✖
        </span>
      </Button>
      <Button onPress={useCallback(() => onButtonPress(0), [onButtonPress])}>
        0
      </Button>
      <Button onPress={onBackspace}>
        <span role="img" aria-label="backspace">
          ⌫
        </span>
      </Button>
    </NumberPadContainer>
  );
}
