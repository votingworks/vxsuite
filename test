#!/usr/bin/env bun

import { readFile } from 'fs/promises';
import { join } from 'path';

const { $ } = Bun;
export {};

const workspacePackageJson = await readFile(
  join(__dirname, 'package.json'),
  'utf8'
);
const { workspaces } = JSON.parse(workspacePackageJson);
const WIDTH = 100;

const results: { pkgPath: string; status: 'success' | 'failure' }[] = [];
let totalPassed = 0;
let totalFailed = 0;

for (const pkgPath of workspaces) {
  const tests = box(
    (async function* () {
      yield `→ bun run test --cwd ${pkgPath}`;
      const bunTest = $`bun run test --cwd ${pkgPath}`.quiet();

      const bunTestResult = await bunTest;

      yield* bunTestResult.stderr.toString('utf-8').split('\n');
      yield `${bunTestResult.exitCode === 0 ? '✅' : '❌'} ${pkgPath}`;

      if (bunTestResult.exitCode === 0) {
        results.push({ pkgPath, status: 'success' });
      } else {
        results.push({ pkgPath, status: 'failure' });
      }
    })()
  );

  let passed = 0;
  let failed = 0;

  for await (const line of tests) {
    const passedMatch = line.match(/(\d+) pass/);
    const failedMatch = line.match(/(\d+) fail/);

    if (passedMatch) {
      passed = parseInt(passedMatch[1], 10);
    }

    if (failedMatch) {
      failed = parseInt(failedMatch[1], 10);
    }

    console.log(line);
  }

  totalPassed += passed;
  totalFailed += failed;
}

console.log(
  `\n${totalPassed + totalFailed} tests: ${totalPassed} passed, ${totalFailed} failed`
);

for (const { pkgPath, status } of results) {
  console.log(`${status === 'success' ? '✅' : '❌'} ${pkgPath}`);
}

async function* box(lines: AsyncIterable<string>): AsyncIterable<string> {
  yield* lines;
}
