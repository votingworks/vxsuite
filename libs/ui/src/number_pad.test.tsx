import userEvent from '@testing-library/user-event';
import { fireEvent, render, screen } from '../test/react_testing_library';

import { NumberPad } from './number_pad';

test('click all pad buttons', () => {
  const onPress = jest.fn();
  const onBackspace = jest.fn();
  const onClear = jest.fn();
  render(
    <NumberPad
      onButtonPress={onPress}
      onBackspace={onBackspace}
      onClear={onClear}
    />
  );
  for (let digit = 0; digit <= 9; digit += 1) {
    userEvent.click(screen.getButton(`${digit}`));
    expect(onPress).toHaveBeenCalledWith(digit);
  }
  expect(onPress).toHaveBeenCalledTimes(10);

  const backspaceButton = screen.getButton('backspace');
  userEvent.click(backspaceButton);
  expect(onBackspace).toHaveBeenCalledTimes(1);

  const clearButton = screen.getButton('clear');
  userEvent.click(clearButton);
  expect(onClear).toHaveBeenCalledTimes(1);
});

// FIXME: It'd be great to use `userEvent.keyboard`, but it doesn't work.
// Maybe because the focused element is a `div`? But if it has a tab index
// then it should work? ¯\_(ツ)_/¯
//
//   userEvent.keyboard('0123456789')
//
function sendKey(key: string): void {
  const charCode = key === 'Backspace' ? undefined : key.charCodeAt(0);
  fireEvent.keyDown(document.activeElement ?? document.body, {
    key,
    charCode,
  });
  if (charCode !== undefined) {
    fireEvent.keyPress(document.activeElement ?? document.body, {
      key,
      charCode,
    });
  }
  fireEvent.keyUp(document.activeElement ?? document.body, {
    key,
    charCode,
  });
}

test('keyboard interaction', () => {
  const onPress = jest.fn();
  const onBackspace = jest.fn();
  const onClear = jest.fn();
  const { container } = render(
    <NumberPad
      onButtonPress={onPress}
      onBackspace={onBackspace}
      onClear={onClear}
    />
  );
  container.focus();

  // some keys should be ignored
  sendKey('z');
  expect(onPress).not.toHaveBeenCalled();
  expect(onBackspace).not.toHaveBeenCalled();
  expect(onClear).not.toHaveBeenCalled();

  for (let digit = 0; digit <= 9; digit += 1) {
    sendKey(`${digit}`);
  }

  for (let digit = 0; digit <= 9; digit += 1) {
    expect(onPress).toHaveBeenCalledWith(digit);
  }

  expect(onPress).toHaveBeenCalledTimes(10);

  sendKey('Backspace');
  expect(onBackspace).toHaveBeenCalledTimes(1);

  sendKey('x');
  expect(onClear).toHaveBeenCalledTimes(1);

  sendKey('Enter');
  // nothing should happen
  expect(onClear).toHaveBeenCalledTimes(1);
  expect(onBackspace).toHaveBeenCalledTimes(1);
  expect(onPress).toHaveBeenCalledTimes(10);
});

test('keyboard interaction when onEnter is defined', () => {
  const onPress = jest.fn();
  const onBackspace = jest.fn();
  const onClear = jest.fn();
  const onEnter = jest.fn();
  const { container } = render(
    <NumberPad
      onButtonPress={onPress}
      onBackspace={onBackspace}
      onClear={onClear}
      onEnter={onEnter}
    />
  );
  container.focus();

  // some keys should be ignored
  sendKey('z');
  expect(onPress).not.toHaveBeenCalled();
  expect(onBackspace).not.toHaveBeenCalled();
  expect(onClear).not.toHaveBeenCalled();
  expect(onEnter).not.toHaveBeenCalled();

  for (let digit = 0; digit <= 9; digit += 1) {
    sendKey(`${digit}`);
  }

  for (let digit = 0; digit <= 9; digit += 1) {
    expect(onPress).toHaveBeenCalledWith(digit);
  }

  expect(onPress).toHaveBeenCalledTimes(10);

  sendKey('Backspace');
  expect(onBackspace).toHaveBeenCalledTimes(1);

  sendKey('x');
  expect(onClear).toHaveBeenCalledTimes(1);

  sendKey('Enter');
  expect(onEnter).toHaveBeenCalledTimes(1);
  expect(onClear).toHaveBeenCalledTimes(1);
  expect(onBackspace).toHaveBeenCalledTimes(1);
  expect(onPress).toHaveBeenCalledTimes(10);
});

test('disabled state', () => {
  const onBackspace = jest.fn();
  const onButtonPress = jest.fn();
  const onClear = jest.fn();
  const onEnter = jest.fn();
  const { container } = render(
    <NumberPad
      disabled
      onBackspace={onBackspace}
      onButtonPress={onButtonPress}
      onClear={onClear}
      onEnter={onEnter}
    />
  );
  container.focus();

  for (let digit = 0; digit <= 9; digit += 1) {
    sendKey(`${digit}`);
  }
  sendKey('Backspace');
  sendKey('x');
  sendKey('Enter');

  expect(onBackspace).not.toHaveBeenCalled();
  expect(onButtonPress).not.toHaveBeenCalled();
  expect(onClear).not.toHaveBeenCalled();
  expect(onEnter).not.toHaveBeenCalled();
});
