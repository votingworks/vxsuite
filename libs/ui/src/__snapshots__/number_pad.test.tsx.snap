// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="sc-eCImvq lhLApJ"
  tabindex="0"
>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    1
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    2
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    3
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    4
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    5
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    6
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    7
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    8
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    9
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    <span
      aria-label="clear"
      role="img"
    >
      ✖
    </span>
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    0
  </button>
  <button
    class="sc-gsDJrp dmKzwC"
    type="button"
  >
    <span
      aria-label="backspace"
      role="img"
    >
      ⌫
    </span>
  </button>
</div>
`;
